src/rate_limiter/src/lib.rs

```diff
@@ -357,23 +374,40 @@ impl RateLimiter {
             TokenType::Ops => self.ops.as_mut(),
         };
         // Try to consume from the token bucket.
-        let success = match token_bucket {
-            Some(bucket) => bucket.reduce(tokens),
+        match token_bucket {
+            Some(bucket) => {
+                let refill_time = bucket.refill_time_ms();
+                match bucket.reduce(tokens) {
+                    // When we report budget is over, there will be no further calls here,
+                    // register a timer to replenish the bucket and resume processing;
+                    // make sure there is only one running timer for this limiter.
+                    BucketReduction::Failure => {
+                        self.activate_timer(TIMER_REFILL_STATE);
+                        false
+                    }
+                    // The operation succeeded and further calls can be made.
+                    BucketReduction::Success => true,
+                    // The operation succeeded as the tokens have been consumed
+                    // but the timer still needs to be armed.
+                    BucketReduction::OverConsumption(ratio) => {
+                        // The operation "borrowed" a number of tokens `ratio` times
+                        // greater than the size of the bucket, and since it takes
+                        // `refill_time` milliseconds to fill an empty bucket, in
+                        // order to enforce the bandwidth limit we need to prevent
+                        // further calls to the rate limiter for
+                        // `ratio * refill_time` milliseconds.
+                        self.activate_timer(TimerState::Oneshot(Duration::from_millis(
+                            (ratio * refill_time as f64) as u64,
+                        )));
+                        true
```

Comment:
There another conceptual problem with this fix.. :

If **all** your packets/consume-requests are larger than bucket size, they **all** go through with _no rate-limiting_ whatsoever. This happens because you allow them through but never block on the limiter. The current logic relies on the idea that a future packet with size < bucket.size will block the limiter.

Reply:
Nevermind, we also expose the `is_blocked()` api which reports the rate-limiter as blocked if there is a timer active, so above statement is false.